requests-for-comments/2025-03-14-dw-testing-production-data.md

```diff
@@ -0,0 +1,89 @@
+   `posthog-s3-datawarehouse-us-east-1` into an ephemeral s3 bucket from a list of deltalake tables from prod → dev
+   environment. Using this approach we can avoid looking at credentials and our Postgres instance. This also means that
+   we have very minimal interactions with production as a whole.
+
+2. Reconstruct stateful instances from the delta tables in our CI flow. This would require different processes for
+   various providers. For example, for MySql we could take an official image, deploy it with ArgoCD, then have a program
+   move the data that has been temporarily copied into our dev s3 bucket into entities we consider external data
+   sources.
+
+3. Run an on demand temporal workflow to test our standard syncing operation as if it’s any other production workflow.
+
+4. The results of these workflows will end up back in s3 via delta as an output. We should be able to run queries at
+   this point.
+
+5. Compare the results of the original copied s3 payloads to the s3 payloads that went through the temporal processor
+   and were synced. This is mostly a parity check to ensure that our syncing process is working as expected.
+
+6. Run test case queries on data that has moved from s3 → external_data_source → s3 either with custom queries or test
+   queries inherited from production.
+
+7. Reverse of #1 to clean up all fragments of delta tables that have been copied into the dev environment and buckets
+   that have. Cleanup fragments in Clickhouse that have been generated by this process as well.
+
+[Data Warehouse Testing Production Flow](./images/dw-testing-production-flo.png)
+
+## Out of Scope
+- Temporal Scheduling Logic - We want to ensure that sync works, not that temporal invokes a schedule at the correct
+  interval.
+- Load testing - For massive queries that exceed thresholds encountered in production. This is mostly meant to test if
+  the schemas and queries for syncs are working, not if we can handle massive transactions.
+
+## Tooling
+- ArgoCd
+    - Source containers (Postgres, MySql, Snowflake)
+    - Source destinations (Postgres, MySql, Snowflake)
+    - Clickhouse (Reuse in dev environment as query layer)
+    - Temporal Runner(Reuse in dev environment as query layer)
+- S3
+- GitHub Actions (Invoke when changing subset of our directories, triggering agent)
+
+## Concerns/Open Questions
+- How feasible is it to recreate instances of more hosted type databases, like Snowflake, would something like
+  LocalStack be an option for us?
+- It is too much work to do this for some many of our integrations, would we become test automation specialists instead
+  of software engineers? How can we maintain this balance over time added more and more source types?
+- What would be the best way to simulate other sources like Salesforce?
+- Should we use a test account for something like Stripe?
+- Would we like to test out complicated joins on queries between multiple sources and events? 
+
```

Comment:
Have you considered any alternative approaches?

For example, maybe a way to compare the data we have in PostHog vs the source is to occasionally run some reconciliation queries, which change depending on the source type. Eg for BigQuery you could run some queries to get row counts in the source vs PostHog, to check breakdowns by a certain column, etc. Then we wouldn't need to spin up our own BigQuery instance.

Reply:
I suppose it depends on what the queries are. If they're just running aggregations like row counts then the actual data egress would be minimal. However it would obviously incur additional compute on the user's end.

Though maybe we could mitigate this:
- choosing a random source to perform the test on, so that each source would only likely have an additional query per week or something
- have an opt-in check box which would be something like "allow us to perform a data integrity check on your data once a week"